src/Tools/dotnet-format/CodeFormatter.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.MSBuild;
+using Microsoft.CodeAnalysis.Formatting;
+using Roslyn.Utilities;
+using CSharpSyntaxKind = Microsoft.CodeAnalysis.CSharp.SyntaxKind;
+using VisualBasicSyntaxKind = Microsoft.CodeAnalysis.VisualBasic.SyntaxKind;
+using Microsoft.CodeAnalysis.Tools.CodeFormatter.MSBuild;
+using System.Diagnostics;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.CodeAnalysis.Tools.CodeFormatter
+{
+    internal class CodeFormatter
+    {
+        public static async Task<int> FormatWorkspaceAsync(ILogger logger, string workspacePath, bool isSolution, CancellationToken token)
+        {
+            logger.LogInformation(string.Format(Resources.Info_FormattingFilesInWorkspace, Path.GetFileName(workspacePath)));
+
+            logger.LogTrace($"Loading workspace.");
+
+            var stopwatch = Stopwatch.StartNew();
+
+            var properties = new Dictionary<string, string>
+            {
+                // This property ensures that XAML files will be compiled in the current AppDomain
+                // rather than a separate one. Any tasks isolated in AppDomains or tasks that create
+                // AppDomains will likely not work due to https://github.com/Microsoft/MSBuildLocator/issues/16.
+                { "AlwaysCompileMarkupFilesInSeparateDomain", bool.FalseString },
+
+                // Use the latest language version to force the full set of available analyzers to run on the project.
+                { "LangVersion", "latest" },
+            };
+            using (var workspace = MSBuildWorkspace.Create(properties))
```

Comment:
I think we want to attach to the `workspace.WorkspaceFailed` event so that the user is notified why a project may have failed to load.